File: legacy/backend/app/api/routes/attributes.py
from fastapi import APIRouter, Depends, HTTPException, Query

from backend.app.api.deps import get_app_container
from backend.app.api.schemas.attributes import (
    AttributeSchemaResponse,
    AttributeSchemaUpsertRequest,
    AttributeValueResponse,
    AttributeValueUpsertRequest,
)
from backend.app.core.container import AppContainer
from backend.app.domain.models.attribute import AttributeSchema, AttributeValue

router = APIRouter(prefix="/api/attributes", tags=["attributes"])


@router.get("/schemas", response_model=list[AttributeSchemaResponse])
def list_attribute_schemas(
    scope: str | None = Query(default=None),
    container: AppContainer = Depends(get_app_container),
):
    return [AttributeSchemaResponse.from_domain(item) for item in container.attribute_queries.list_schemas(scope=scope)]


@router.post("/schemas", response_model=AttributeSchemaResponse)
def create_attribute_schema(
    payload: AttributeSchemaUpsertRequest,
    container: AppContainer = Depends(get_app_container),
):
    created = container.attribute_commands.create_schema(
        AttributeSchema(
            id="new",
            scope=payload.scope,
            key=payload.key,
            label=payload.label,
            value_type=payload.value_type,
            description=payload.description,
            default_value=payload.default_value,
            enum_options=payload.enum_options,
            display_policy=payload.display_policy,
            access_policy=payload.access_policy,
            mutation_policy=payload.mutation_policy,
            influence_policy=payload.influence_policy,
            projection_policy=payload.projection_policy,
        )
    )
    return AttributeSchemaResponse.from_domain(created)


@router.put("/schemas/{schema_id}", response_model=AttributeSchemaResponse)
def update_attribute_schema(
    schema_id: str,
    payload: AttributeSchemaUpsertRequest,
    container: AppContainer = Depends(get_app_container),
):
    updated = container.attribute_commands.update_schema(
        schema_id,
        AttributeSchema(
            id=schema_id,
            scope=payload.scope,
            key=payload.key,
            label=payload.label,
            value_type=payload.value_type,
            description=payload.description,
            default_value=payload.default_value,
            enum_options=payload.enum_options,
            display_policy=payload.display_policy,
            access_policy=payload.access_policy,
            mutation_policy=payload.mutation_policy,
            influence_policy=payload.influence_policy,
            projection_policy=payload.projection_policy,
        ),
    )
    if updated is None:
        raise HTTPException(status_code=404, detail="Attribute schema not found")
    return AttributeSchemaResponse.from_domain(updated)


@router.get("/values", response_model=list[AttributeValueResponse])
def list_attribute_values(
    owner_type: str | None = Query(default=None),
    owner_id: str | None = Query(default=None),
    schema_id: str | None = Query(default=None),
    container: AppContainer = Depends(get_app_container),
):
    return [
        AttributeValueResponse.from_domain(item)
        for item in container.attribute_queries.list_values(
            owner_type=owner_type,
            owner_id=owner_id,
            schema_id=schema_id,
        )
    ]


@router.put("/values", response_model=AttributeValueResponse)
def upsert_attribute_value(
    payload: AttributeValueUpsertRequest,
    container: AppContainer = Depends(get_app_container),
):
    value = container.attribute_commands.upsert_value(
        AttributeValue(
            id="new",
            schema_id=payload.schema_id,
            owner_type=payload.owner_type,
            owner_id=payload.owner_id,
            value=payload.value,
            source=payload.source,
        )
    )
    return AttributeValueResponse.from_domain(value)

File: src-tauri/src/commands/attributes.rs
use crate::models::attribute::*;
use crate::state::AppState;
use tauri::State;

#[tauri::command]
pub async fn list_attribute_schemas(
    state: State<'_, AppState>,
    scope: Option<String>,
) -> Result<Vec<AttributeSchema>, String> {
    let db = state.db.lock().await;
    let repo = crate::db::repositories::attribute_repo::AttributeRepository::new(db.conn());
    repo.list_schemas(scope.as_deref())
}

#[tauri::command]
pub async fn create_attribute_schema(
    state: State<'_, AppState>,
    request: AttributeSchemaCreateRequest,
) -> Result<AttributeSchema, String> {
    let db = state.db.lock().await;
    let repo = crate::db::repositories::attribute_repo::AttributeRepository::new(db.conn());
    repo.create_schema(&request)
}

#[tauri::command]
pub async fn update_attribute_schema(
    state: State<'_, AppState>,
    id: String,
    request: AttributeSchemaCreateRequest,
) -> Result<AttributeSchema, String> {
    let db = state.db.lock().await;
    let repo = crate::db::repositories::attribute_repo::AttributeRepository::new(db.conn());
    repo.update_schema(&id, &request)
}

#[tauri::command]
pub async fn list_attribute_values(
    state: State<'_, AppState>,
    owner_type: Option<String>,
    owner_id: Option<String>,
    schema_id: Option<String>,
) -> Result<Vec<AttributeValue>, String> {
    let db = state.db.lock().await;
    let repo = crate::db::repositories::attribute_repo::AttributeRepository::new(db.conn());
    repo.list_values(
        owner_type.as_deref(),
        owner_id.as_deref(),
        schema_id.as_deref(),
    )
}

#[tauri::command]
pub async fn upsert_attribute_value(
    state: State<'_, AppState>,
    request: AttributeValueUpsertRequest,
) -> Result<AttributeValue, String> {
    let db = state.db.lock().await;
    let repo = crate::db::repositories::attribute_repo::AttributeRepository::new(db.conn());
    repo.upsert_value(&request)
}
